{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Python version 3.4.3 |Anaconda 2.3.0 (x86_64)| (default, Mar  6 2015, 12:07:41) \n",
      "[GCC 4.2.1 (Apple Inc. build 5577)]\n",
      "IMPORTANT: IF THE ABOVE SAYS PYTHON 3.X DO NOT USE QUOTES IN RESPONSES.\n",
      " \n",
      "Hi baby, welcome to SynBioCAD. I'm here for you.\n",
      "Right now, we assume you're working in S. cerevisiae.\n",
      " \n",
      "What do you want to do today? Just press the number\n",
      "\"1\" to insert DNA into a characterized locus\n",
      "\"2\" to edit an existing gene, e.g., delete or replace\n",
      "\"3\" to build a custom cassette\n",
      " \n",
      "Your answer: 1\n",
      " \n",
      "          cutChr           cutSequence\n",
      "cutName                               \n",
      "ARS106a   Scer01  ATACGGTCAGGGTAGCGCCC\n",
      "ARS208a   Scer02  GTCCGCTAAACAAAAGATCT\n",
      "ARS208b   Scer02  TAGGATACAAGTGTCTGTGC\n",
      "ARS308a   Scer03  CACTTGTCAAACAGAATATA\n",
      "LEU2a     Scer03  AAAGGTGAGAGCGCCGGAAC\n",
      "MATαa     Scer03  GCCCTGGATAGCGAAGCGTG\n",
      "MATαb     Scer03  AATGATGTCTGGGTTTTGTT\n",
      "MATαc     Scer03  TGGGAACAAGAGCAAGACGA\n",
      "RDS1a     Scer03  ATTCAATACGAAATGTGTGC\n",
      "ARS416a   Scer04  CCTTCAAAGGATTCACCAGA\n",
      "ARS416b   Scer04  TTAGGGACAGTTAGAGGCGG\n",
      "ARS416c   Scer04  CAATTCAGGCCAAATCTTGA\n",
      "ARS416d   Scer04  TAGTGCACTTACCCCACGTT\n",
      "ARS511b   Scer05  CAGTGTATGCCAGTCAGCCA\n",
      "URA3a     Scer05  ATTCGCTATTATTCAAGCTT\n",
      "URA3b     Scer05  CCTTCGTTCTTCCTTCTGCT\n",
      "CAN1y     Scer05  GATACGTTCTCTATGGAGGA\n",
      "ARS607b   Scer06  CCAGCGTAAGGTAAATATTA\n",
      "ARS607c   Scer06  CTATTTTTGCTTTCTGCACA\n",
      "SAP155b   Scer06  GGTTTTCATACTGGGGCCGC\n",
      "SAP155c   Scer06  ATGAAAGACAACTATAGGGC\n",
      "ARS720a   Scer07  CAACAATTGTTACAATAGTA\n",
      "ARS805a   Scer08  TTATTTGAATGATATTTAGT\n",
      "ARS911b   Scer09  GTAATATTGTCTTGTTTCCC\n",
      "ARS1021a  Scer10  AAGTTGGCTAGCCATATTTT\n",
      "ARS1021b  Scer10  CCTCTGTGTGGTGGTAATTG\n",
      "ARS1014a  Scer10  TTATGTGCGTATTGCTTTCA\n",
      "ARS1114a  Scer11  CTTGTGAAACAAATAATTGG\n",
      "ARS1206a  Scer12  CGAACATTTTTCCATGCGCT\n",
      "ARS1309a  Scer13  CCTGTGGTGACTACGTATCC\n",
      "ARS1414a  Scer14  GCGCCACAGTTTCAAGGGTC\n",
      "YOLCd1a   Scer15  CTAGAATTTCCATTTTGCGT\n",
      "YOLCd1b   Scer15  GACTAGTTAAGCGAGCATGT\n",
      "HIS3b     Scer15  AATATAGAGTGTACTAGAGG\n",
      "ARS1622a  Scer16  TAAAGCCACCACATCGCAAA\n",
      "ARS1622b  Scer16  GTCACGTTCCTGAGGTTACT\n",
      "YPRCd15a  Scer16  ATATGTTTGGTTTCGATTGT\n",
      "YPRCd15c  Scer16  AATCCGAACAACAGAGCATA\n",
      "\n",
      "Which cut site do you want, e.g., 208a?\n",
      "Your answer: ARS1622a\n",
      "\n",
      "Found your homology regions\n",
      "\n",
      "What do you want to do to ARS1622a?\n",
      "\"1\" to insert a custom, prebuilt cassette\n",
      "\"2\" to construct a cassette using standard promoter and terminator names, and a custom ORF\n",
      "\"3\" to construct a donor containing two standard cassettes (pointing away from each other)\n",
      " \n",
      "Your answer: 2\n",
      "I'm going to build a standard cassette in which promoter is 600nt, terminator 250nt.\n",
      "First, which PROMOTER do you want to use, e.g., TDH3\n",
      "Your answer: TEF1p\n"
     ]
    },
    {
     "ename": "UnboundLocalError",
     "evalue": "local variable 'GeneSeq' referenced before assignment",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mUnboundLocalError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-3-115ef3f2ecb3>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m    598\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    599\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 600\u001b[0;31m \u001b[0maskUser\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    601\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-3-115ef3f2ecb3>\u001b[0m in \u001b[0;36maskUser\u001b[0;34m()\u001b[0m\n\u001b[1;32m     55\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mAction\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m\"1\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     56\u001b[0m         \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\" \"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 57\u001b[0;31m         \u001b[0meditEmpty\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     58\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     59\u001b[0m     \u001b[0;32melif\u001b[0m \u001b[0mAction\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m\"2\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-3-115ef3f2ecb3>\u001b[0m in \u001b[0;36meditEmpty\u001b[0;34m()\u001b[0m\n\u001b[1;32m    131\u001b[0m     \u001b[0;32melif\u001b[0m \u001b[0mtypeEdit\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m\"2\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    132\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 133\u001b[0;31m         \u001b[0mPromoterRec\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0morfRecord\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mTerminatorRec\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mbuildCassette\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    134\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    135\u001b[0m         \u001b[0mfragments\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mUpHomRec\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mPromoterRec\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0morfRecord\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mTerminatorRec\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mDownHomRec\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-3-115ef3f2ecb3>\u001b[0m in \u001b[0;36mbuildCassette\u001b[0;34m()\u001b[0m\n\u001b[1;32m    450\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    451\u001b[0m     \u001b[0mPromoterName\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0minput\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Your answer: \"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 452\u001b[0;31m     \u001b[0mPromoterGeneRec\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mfetchGene\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mPromoterName\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    453\u001b[0m     \u001b[0mPromoterRec\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mfetchNeighbor\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mPromoterGeneRec\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\"upstream\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;36m600\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    454\u001b[0m     \u001b[0mPromoterRec\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mPromoterRec\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0;34m\"ps\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-3-115ef3f2ecb3>\u001b[0m in \u001b[0;36mfetchGene\u001b[0;34m(GeneName)\u001b[0m\n\u001b[1;32m    286\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    287\u001b[0m     \u001b[0;31m#let's create a record for the oldGene\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 288\u001b[0;31m     \u001b[0mGeneRecord\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mSeqRecord\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mGeneSeq\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mid\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mGeneSysName\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    289\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    290\u001b[0m     \u001b[0;31m#now let's add some more information to make it useful\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mUnboundLocalError\u001b[0m: local variable 'GeneSeq' referenced before assignment"
     ]
    }
   ],
   "source": [
    "'''\n",
    "CustomCodon\n",
    "by Leo d'Espaux and William Zhuang\n",
    "\n",
    "Current as of 6/24/15\n",
    "\n",
    "Recent updates:\n",
    "*Stitch multiple random fragments\n",
    "*yet to fix name for the above\n",
    "'''\n",
    "\n",
    "\n",
    "# import libraries we're using\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio import SeqIO\n",
    "from Bio.SeqUtils import MeltingTemp\n",
    "import copy\n",
    "\n",
    "from intermine.webservice import Service\n",
    "\n",
    "from pandas import *\n",
    "from pandas import DataFrame, read_csv\n",
    "import pandas as pd #this is how I usually import pandas\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "import sys\n",
    "\n",
    "\n",
    "            \n",
    "CodonTable = pd.read_excel(\"PytoYali.xlsx\", index_col=\"Codon\")\n",
    "    \n",
    "print(CodonTable)\n",
    "    \n",
    "print(\"\")\n",
    "print(\"What's your CDS?\")\n",
    "OriginalSeq=Seq(input(\"Your answer: \"))\n",
    "\n",
    "FinalSeq=Seq(\"ATG\")\n",
    "\n",
    "length=len(OriginalSeq)/3-1\n",
    "\n",
    "for i in range(len(OriginalSeq)/3-1):\n",
    "    \n",
    "\n",
    "\n",
    "    # cutFrame has the 20nt sequence of the gRNA saved, which we load\n",
    "    cutSequence=cutFrame.loc[cutName,'cutSequence']\n",
    "\n",
    "    # We need to load the chromosome where \n",
    "    location=cutFrame.loc[cutName,'cutChr']+\".fasta\"\n",
    "     \n",
    "    ChromosomeSeq=SeqIO.read(location, \"fasta\").seq\n",
    "    \n",
    "    if ChromosomeSeq.find(cutSequence)==-1:\n",
    "        ChromosomeSeq=ChromosomeSeq.reverse_complement()\n",
    "        if ChromosomeSeq.find(cutSequence)==-1:\n",
    "            print(\"CAN'T FIND THE CUT SITE IN YOUR SEQUENCE\")\n",
    "\n",
    "    StartIndex=ChromosomeSeq.find(cutSequence)\n",
    "    EndIndex=StartIndex+34\n",
    "    \n",
    "\n",
    "    \n",
    "    UpSeq=ChromosomeSeq[StartIndex-HomologyLength:StartIndex]\n",
    "    DownSeq=ChromosomeSeq[EndIndex:EndIndex+HomologyLength]\n",
    "        \n",
    "    UpHomRec = SeqRecord(UpSeq, name=cutName)\n",
    "    DownHomRec = SeqRecord(DownSeq, name=cutName)\n",
    "    \n",
    "    print(\"\")\n",
    "    print(\"Found your homology regions\")\n",
    "    print(\"\")\n",
    "\n",
    "              \n",
    "    print(\"What do you want to do to \" + cutName + \"?\")\n",
    "    print(\"\\\"1\\\" to insert a custom, prebuilt cassette\")\n",
    "    print(\"\\\"2\\\" to construct a cassette using standard promoter and terminator names, and a custom ORF\")\n",
    "    print(\"\\\"3\\\" to construct a donor containing two standard cassettes (pointing away from each other)\")\n",
    "    print(\" \")\n",
    "\n",
    "    typeEdit=input(\"Your answer: \")\n",
    "    if typeEdit==\"1\":\n",
    "    \n",
    "        orfName=input(\"What's the name of your custom gene or cassette?\")\n",
    "        orfSeq=Seq(input(\"What's the sequence?\"))\n",
    "        \n",
    "        orfRecord=SeqRecord(orfSeq,name=orfName)\n",
    "        \n",
    "        fragments=[UpHomRec, orfRecord, DownHomRec]\n",
    "\n",
    "    \n",
    "    elif typeEdit == \"2\":\n",
    "        \n",
    "        PromoterRec, orfRecord, TerminatorRec = buildCassette()\n",
    "        \n",
    "        fragments=[UpHomRec, PromoterRec, orfRecord, TerminatorRec, DownHomRec]\n",
    "\n",
    "    elif typeEdit == \"3\":\n",
    "        \n",
    "        print(\"Your construct will look like this: \")\n",
    "        print(\"up hom-<terminator1-gene1-promoter1<->promoter2-gene2-terminator2>-downhom\")\n",
    "        print(\"OK, let's build your FIRST cassette\")\n",
    "        PromoterRec1, orfRecord1, TerminatorRec1 = buildCassette()\n",
    "         \n",
    "        #now in the construction, these will be in the antisense direction, so let's change their sequences\n",
    "        rTerminatorRec1=flipRecord(TerminatorRec1)\n",
    "        rorfRecord1=flipRecord(orfRecord1)\n",
    "        rPromoterRec1=flipRecord(PromoterRec1)\n",
    "        \n",
    "        print(\"OK, let's build your SECOND cassette\")\n",
    "        PromoterRec2, orfRecord2, TerminatorRec2 = buildCassette()\n",
    "\n",
    "        fragments=[UpHomRec, rTerminatorRec1, rorfRecord1, rPromoterRec1,PromoterRec2,orfRecord2, TerminatorRec2,DownHomRec]\n",
    "\n",
    "    stitch(fragments)\n",
    "    \n",
    "    \n",
    "    \n",
    "def editExisting():\n",
    "    \n",
    "    print(\" \")\n",
    "    print(\"Which locus do you want to edit? Tell me a common name, e.g., \\\"OAF1\\\": \")\n",
    "    print(\"I'm smart and pretty and I can fetch it for you.\")\n",
    "    \n",
    "    GeneName= input(\"Your answer: \")\n",
    "    \n",
    "    print(\"OK let me fetch that for you...\")\n",
    "    print(\"\")\n",
    "\n",
    "        \n",
    "    OrigGeneRecord=fetchGene(GeneName)\n",
    "    #note that this returns a seqrecord\n",
    "    \n",
    "    print(\"\")\n",
    "    print(\"I found \"+OrigGeneRecord.features+\":\")\n",
    "    print(OrigGeneRecord.description)\n",
    "\n",
    "\n",
    "    # We make seqrecords since that's what we carry through later in the program\n",
    "    UpHomRec = fetchNeighbor(OrigGeneRecord, \"upstream\", HomologyLength )\n",
    "    DownHomRec = fetchNeighbor(OrigGeneRecord, \"downstream\", HomologyLength )\n",
    "    \n",
    "    #print(\"OK I found them.\")\n",
    "\n",
    "    #print(\" \")\n",
    "    #print(\"UpstreamHomology reghhion is\")\n",
    "    #print(UpHomRec.seq)\n",
    "    \n",
    "    #print(\" \")\n",
    "    #print(\"DownstreamHomology reghhion is\")\n",
    "    #print(DownHomRec.seq)\n",
    "    \n",
    "    print(\" \")\n",
    "    print(\"What do you want to do to this gene?\")\n",
    "    print(\"\\\"1\\\" delete the protein coding DNA sequence (CDS)\")\n",
    "    print(\"\\\"2\\\" replace the CDS with another CDS, or other pre-made fragment\")\n",
    "    print(\"\\\"3\\\" replace the CDS with a standard cassette I will build for you\")\n",
    "    print(\"\\\"4\\\" replace the CDS with a custom cassette (soon)\")\n",
    "    print(\"\\\"5\\\" replace a specified region near your target gene (soon)\")\n",
    "    \n",
    "    print(\" \")\n",
    "    Action=input(\"Your answer: \")\n",
    "                        \n",
    "    #note that in all the below, we want to have fragments be records\n",
    "    if Action==\"1\":\n",
    "        fragments=[UpHomRec,DownHomRec]\n",
    "    \n",
    "    if Action==\"2\":\n",
    "        print(\" \")\n",
    "        NewGeneName=input(\"What's the name of the gene you're inserting?\")\n",
    "        NewGeneSeq=Seq(input(\"What's the sequence of your new gene?\"))\n",
    "        InsertRec = SeqRecord(NewGeneSeq, name=NewGeneName)\n",
    "        fragments=[UpHomRec, InsertRec, DownHomRec]\n",
    "\n",
    "    if Action==\"3\":\n",
    "        print(\" \")\n",
    "        print(\"I got you baby. I don't want you to strain your eyes looking up stuff.\")\n",
    "        print(\" \")\n",
    "\n",
    "        PromoterRec, orfRecord, TerminatorRec = standardCassette()\n",
    "        fragments=[UpHomRec, PromoterRec, orfRecord, TerminatorRec, DownHomRec] #we need to finish buildcassette to add InsertRec here\n",
    "\n",
    "    if Action==\"4\":\n",
    "        print(\"How many pieces (other than homology fragments) are you stitching together.\")\n",
    "        Npieces=input(\"Your answer: \")\n",
    "        output = standardCassette()\n",
    "\n",
    "        fragments=[UpHomRec, DownHomRec] #we need to finish buildcassette to add InsertRec here\n",
    "\n",
    "        \n",
    "    stitch(fragments)\n",
    "\n",
    "    \n",
    "    \n",
    "def buildCustom():\n",
    "    print(\"How do you want your cassettes?\")\n",
    "    print(\"\\\"1\\\" to build a cassette with N pieces and no variants.\")\n",
    "    print(\"\\\"2\\\" to build a cassette with N pieces and one variant site.\")\n",
    "\n",
    "    answer = input(\"Your answer:\")\n",
    "\n",
    "    if answer == \"1\":\n",
    "        N = int(input(\"How many pieces (N) in your custom cassette:\"))\n",
    "        fragments = variableCassette(N)[0]\n",
    "        stitch(fragments)\n",
    "\n",
    "    elif answer == \"2\":\n",
    "        N = int(input(\"How many pieces (N) are in your custom cassette: \"))\n",
    "        toVary = int(input(\"Which piece would you like to vary: \"))\n",
    "        variants = int(input(\"How many variants would you like: \"))\n",
    "        fragsList = variableCassette(N, toVary, variants)\n",
    "        # Make the entire thing first. Then just primers for the variant.\n",
    "        print(\"\")\n",
    "        print(\"First we'll get primers for the entire cassette\")\n",
    "        stitch(fragsList[0])\n",
    "        print(\"-------------------------------------------------\")\n",
    "        print(\"Now we'll get primers just for the variants\")\n",
    "\n",
    "        for n in range(N-1):\n",
    "            print(\"For variant \" + str(n) + \":\")\n",
    "            frags = fragsList[n + 1][toVary - 2: toVary + 1]\n",
    "            stitch(frags)\n",
    "            \n",
    "            \n",
    "    #------------------------------ FETCH FUNCTIONS -------------------------------------\n",
    "\n",
    "    \n",
    "def fetchGene(GeneName):\n",
    "    \n",
    "    service = Service(\"http://yeastmine.yeastgenome.org/yeastmine/service\")\n",
    "    template = service.get_template('Gene_GenomicDNA')\n",
    "\n",
    "    rows = template.rows(\n",
    "        E = {\"op\": \"LOOKUP\", \"value\": GeneName, \"extra_value\": \"S. cerevisiae\"}\n",
    "    )\n",
    "    \n",
    "    # this service seems to return multiple similar genes but we want the first one only, so count\n",
    "    # and it returns information about the gene you want\n",
    "    count=0\n",
    "    for row in rows:\n",
    "        \n",
    "        count=count+1\n",
    "        if count==1:\n",
    "            descr= row[\"description\"]\n",
    "            GeneSeq=Seq(row[\"sequence.residues\"])\n",
    "            GeneSysName=row[\"secondaryIdentifier\"]\n",
    "       \n",
    "    #let's create a record for the oldGene\n",
    "    GeneRecord = SeqRecord(GeneSeq, id=GeneSysName)\n",
    "    \n",
    "    #now let's add some more information to make it useful\n",
    "    GeneRecord.name=GeneName\n",
    "    GeneRecord.features=GeneSysName\n",
    "    GeneRecord.description=descr\n",
    "\n",
    "    return GeneRecord \n",
    "       \n",
    "    \n",
    "    \n",
    "def fetchNeighbor(NeighborRecord, direction, distance):\n",
    "\n",
    "    # let's load the appropriate chromosome file. The record of the gene we looked up\n",
    "    # contains in the \"features\" the systematic name, wherein the second letter\n",
    "    # corresponds to chromosome number, e.g., 1=A etc\n",
    "    if NeighborRecord.features[1]==\"A\":\n",
    "        ChromosomeRec=SeqIO.read(\"Scer01.fasta\", \"fasta\")\n",
    "    if NeighborRecord.features[1]==\"B\":\n",
    "        ChromosomeRec=SeqIO.read(\"Scer02.fasta\", \"fasta\")\n",
    "    if NeighborRecord.features[1]==\"C\":\n",
    "        ChromosomeRec=SeqIO.read(\"Scer03.fasta\", \"fasta\")\n",
    "    if NeighborRecord.features[1]==\"D\":\n",
    "        ChromosomeRec=SeqIO.read(\"Scer04.fasta\", \"fasta\")\n",
    "    if NeighborRecord.features[1]==\"E\":\n",
    "        ChromosomeRec=SeqIO.read(\"Scer05.fasta\", \"fasta\")\n",
    "    if NeighborRecord.features[1]==\"F\":\n",
    "        ChromosomeRec=SeqIO.read(\"Scer06.fasta\", \"fasta\")\n",
    "    if NeighborRecord.features[1]==\"G\":\n",
    "        ChromosomeRec=SeqIO.read(\"Scer07.fasta\", \"fasta\")\n",
    "    if NeighborRecord.features[1]==\"H\":\n",
    "        ChromosomeRec=SeqIO.read(\"Scer08.fasta\", \"fasta\")\n",
    "    if NeighborRecord.features[1]==\"I\":\n",
    "        ChromosomeRec=SeqIO.read(\"Scer09.fasta\", \"fasta\")\n",
    "    if NeighborRecord.features[1]==\"J\":\n",
    "        ChromosomeRec=SeqIO.read(\"Scer10.fasta\", \"fasta\")\n",
    "    if NeighborRecord.features[1]==\"K\":\n",
    "        ChromosomeRec=SeqIO.read(\"Scer11.fasta\", \"fasta\")\n",
    "    if NeighborRecord.features[1]==\"L\":\n",
    "        ChromosomeRec=SeqIO.read(\"Scer12.fasta\", \"fasta\")\n",
    "    if NeighborRecord.features[1]==\"M\":\n",
    "        ChromosomeRec=SeqIO.read(\"Scer13.fasta\", \"fasta\")\n",
    "    if NeighborRecord.features[1]==\"N\":\n",
    "        ChromosomeRec=SeqIO.read(\"Scer14.fasta\", \"fasta\")\n",
    "    if NeighborRecord.features[1]==\"O\":\n",
    "        ChromosomeRec=SeqIO.read(\"Scer15.fasta\", \"fasta\")\n",
    "    if NeighborRecord.features[1]==\"P\":\n",
    "        ChromosomeRec=SeqIO.read(\"Scer16.fasta\", \"fasta\") \n",
    "\n",
    "    \n",
    "    \n",
    "    # let's explicitely name the sequences from the seq record\n",
    "    NeighborSeq=NeighborRecord.seq\n",
    "    ChromosomeSeq=ChromosomeRec.seq\n",
    "    \n",
    "    # flip the sequence to orient with respect to the old gene\n",
    "    if ChromosomeSeq.find(NeighborSeq)==-1:\n",
    "        ChromosomeSeq=ChromosomeSeq.reverse_complement()\n",
    "\n",
    "    StartIndex=ChromosomeSeq.find(NeighborSeq)\n",
    "    EndIndex=StartIndex+len(NeighborSeq)\n",
    "    \n",
    "    if direction==\"upstream\":\n",
    "        DesiredSeq=ChromosomeSeq[StartIndex-distance:StartIndex]\n",
    "    if direction==\"downstream\":\n",
    "        DesiredSeq=ChromosomeSeq[EndIndex:EndIndex+distance]\n",
    "\n",
    "    \n",
    "    \n",
    "    \n",
    "    NeighborRec = SeqRecord(DesiredSeq, name=NeighborRecord.name)\n",
    "    \n",
    "    return NeighborRec\n",
    "\n",
    "    #print(NeighborRec)\n",
    "\n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    #------------------------------------ CONSTRUCTING STUFF --------------------------------------\n",
    "\n",
    "def getPrimer(currRecord):\n",
    "    \n",
    "\n",
    "    mp = 0\n",
    "    length = 0\n",
    "    primer = Seq(\"\")\n",
    "\n",
    "    seq=currRecord.seq\n",
    "    \n",
    "    while mp <= PrimerMaxTm and length <= PrimerMaxLen:\n",
    "        primer = primer + seq[length]\n",
    "        mp = MeltingTemp.Tm_staluc(primer)\n",
    "        length += 1\n",
    "\n",
    "    return primer           \n",
    "        \n",
    "def overhangPrimer(currRecord,prevSeq):\n",
    "    #let's get the template-binding primer first\n",
    "    primer=getPrimer(currRecord)\n",
    "    \n",
    "    \n",
    "    #OK let's work on the overhang\n",
    "    maxOhLen=PrimerMaxLen-len(primer)    \n",
    "    maxFrac=1\n",
    "    \n",
    "    #let's decide on a max overhang length\n",
    "    if round(len(primer)*(OverhangMaxFrac+1)) < 60:\n",
    "             maxOhLen=round(len(primer)*OverhangMaxFrac)\n",
    "    \n",
    "    #the index must be an integer!!!\n",
    "    maxOhLen=int(maxOhLen)\n",
    "    ohprimer=prevSeq.seq[-maxOhLen:]+primer #we add the .seq so that it returns a string\n",
    "    \n",
    "    return ohprimer      \n",
    "        \n",
    "def standardCassette():\n",
    "    \n",
    "    #first, the promoter\n",
    "    print(\"I'm going to build a standard cassette in which promoter is 600nt, terminator 250nt.\")\n",
    "    print(\"\")\n",
    "    print(\"Which PROMOTER do you want to use, e.g., TDH3\")\n",
    "    \n",
    "    PromoterName=input(\"Your answer: \")\n",
    "    PromoterGeneRec=fetchGene(PromoterName)\n",
    "    PromoterRec=fetchNeighbor(PromoterGeneRec,\"upstream\",600)\n",
    "    PromoterRec.name=PromoterRec.name+\"ps\"\n",
    "    \n",
    "    \n",
    "    #second, the terminator\n",
    "    print(\"Which TERMINATOR do you want to use, e.g., ADH1\")\n",
    "    TerminatorName = input('Your answer: ')\n",
    "    TerminatorGeneRec=fetchGene(TerminatorName)\n",
    "    TerminatorRec=fetchNeighbor(TerminatorGeneRec,\"downstream\",250)\n",
    "    TerminatorRec.name=TerminatorRec.name+\"ts\"\n",
    "    \n",
    "    \n",
    "    #and last, the gene\n",
    "    print(\"What do you want to call the CDS you're inserting?\")\n",
    "    orfName = input(\"Your answer: \")\n",
    "    \n",
    "    print(\"What's the sequence\")\n",
    "    orfSeq=input(\"Your answer: \")\n",
    "    \n",
    "    orfRecord=SeqRecord(Seq(orfSeq), name=orfName)\n",
    "    \n",
    "    insertRec=[PromoterRec,orfRecord,TerminatorRec]\n",
    "    return PromoterRec, orfRecord, TerminatorRec\n",
    "          \n",
    "def buildCassette():\n",
    "    \n",
    "    #first, the promoter\n",
    "    print(\"I'm going to build a standard cassette in which promoter is 600nt, terminator 250nt.\") \n",
    "    print(\"First, which PROMOTER do you want to use, e.g., TDH3\")\n",
    "    \n",
    "    PromoterName=input(\"Your answer: \")\n",
    "    PromoterGeneRec=fetchGene(PromoterName)\n",
    "    PromoterRec=fetchNeighbor(PromoterGeneRec,\"upstream\",600)\n",
    "    PromoterRec.name=PromoterRec.name+\"ps\"\n",
    "    \n",
    "    \n",
    "    #second, the terminator\n",
    "    print(\"Which TERMINATOR do you want to use, e.g., ADH1\")\n",
    "    TerminatorName = input('Your answer: ')\n",
    "    TerminatorGeneRec=fetchGene(TerminatorName)\n",
    "    TerminatorRec=fetchNeighbor(TerminatorGeneRec,\"downstream\",250)\n",
    "    TerminatorRec.name=TerminatorRec.name+\"ts\"\n",
    "    \n",
    "    #and last, the gene\n",
    "    print(\"What is the name of your gene, e.g., KlGapDH\")\n",
    "    orfName = input(\"Your answer: \")\n",
    "    \n",
    "    print(\"What's the sequence\")\n",
    "    orfSeq=input(\"Your answer: \")\n",
    "    \n",
    "    orfRecord=SeqRecord(Seq(orfSeq), name=orfName)\n",
    "    \n",
    "    insertRec=[PromoterRec,orfRecord,TerminatorRec]\n",
    "    return PromoterRec, orfRecord, TerminatorRec\n",
    "    \n",
    "def variableCassette(N, toVary = 0, variants = 0):\n",
    "    print(\"\")\n",
    "    print(\"Let's start building.\")\n",
    "    print(\"\")\n",
    "    \n",
    "    # Store both name and sequence in a SeqRecord\n",
    "    # Append them to a list\n",
    "    # Return list as fragments to be stitched\n",
    "\n",
    "    records = []\n",
    "    for n in range(N):\n",
    "        name = input(\"What is the name of sequence \" + str(n+1) +\":\")\n",
    "        sequence = input(\"What is the sequence of this fragment:\")\n",
    "        print(\"\")\n",
    "        Rec = SeqRecord(Seq(sequence), id = str(n+1))\n",
    "        Rec.name = name\n",
    "        records.append(Rec)\n",
    "\n",
    "    variantRecords = []\n",
    "    variantRecords.append(records)\n",
    "    # This only happens if there are variants.\n",
    "    if variants > 0:\n",
    "        print(\"Time to make those variants you wanted.\")\n",
    "        for n in range(variants-1):\n",
    "            name = input(\"What is the name of variant \" + str(n+1) + \":\")\n",
    "            sequence = input(\"What is the sequence of this variant:\")\n",
    "            Rec = SeqRecord(Seq(sequence), id = str(n+1))\n",
    "            Rec.name = name\n",
    "            # Make a copy of the original, switch the fragments and add it to the list. \n",
    "            # Deep-copy ensures there are no pointer issues\n",
    "            tempVariant = copy.deepcopy(records)\n",
    "            tempVariant[toVary - 1] = Rec\n",
    "            variantRecords.append(copy.deepcopy(tempVariant))\n",
    "            print(\"\")\n",
    "\n",
    "    # Returns a list of lists of the SeqRecords of the fragments\n",
    "    return variantRecords\n",
    "\n",
    "def stitch(fragments):\n",
    "    #this function takes seq records and prints primers\n",
    "    \n",
    "    #let's make an empty sequence file\n",
    "    Nfrags=len(fragments)\n",
    "    donor=Seq(\"\")\n",
    "    index=[]\n",
    "    print(\"\")\n",
    "    for i in range (0, Nfrags):\n",
    "        donor=donor+fragments[i]\n",
    "    \n",
    "    for i in range (0, Nfrags):\n",
    "        if i==0:\n",
    "            print(\"Lup\"+ fragments[i].name + \" \" + getPrimer(donor))\n",
    "            print(\"Rup\"+ fragments[i].name + \"(\" + fragments[i+1].name + \") \" + overhangPrimer(fragments[i].reverse_complement(),fragments[i+1].reverse_complement()))\n",
    "        elif i==Nfrags-1:\n",
    "            print(\"Ldown\"+ fragments[i].name + \"(\" + fragments[i-1].name + \") \" + overhangPrimer(fragments[i],fragments[i-1]))\n",
    "            print(\"Rdown\"+ fragments[i].name + \" \" + getPrimer(donor.reverse_complement()))\n",
    "        else:\n",
    "            print(\"L\"+ fragments[i].name + \"(\" + fragments[i-1].name + \") \" + overhangPrimer(fragments[i],fragments[i-1]))\n",
    "            print(\"R\"+ fragments[i].name + \"(\" + fragments[i+1].name + \") \" + overhangPrimer(fragments[i].reverse_complement(),fragments[i+1].reverse_complement()))\n",
    "\n",
    "    print(\"\")\n",
    "    print(\"Your donor DNA cassette, has the following bp length and sequence:\")\n",
    "\n",
    "\n",
    "    print(\"\")\n",
    "    print(len(donor.seq))\n",
    "    print(\"\")\n",
    "\n",
    "    print(donor.seq)\n",
    "\n",
    "    print(\"\")\n",
    "    print(\"You might want to copy this entire prompt and save it for your records.\")\n",
    "    \n",
    "        \n",
    "def flipRecord(origRecord):\n",
    "    origRecord.seq=origRecord.seq.reverse_complement()\n",
    "    origRecord.name=\"r\"+origRecord.name\n",
    "    return origRecord\n",
    "    \n",
    "def pickCut(Sequence):\n",
    "    from prettytable import from_csv\n",
    "\n",
    "    print(\"Which of these characterized loci is going to get lucky: \")\n",
    "\n",
    "    fp = open(\"cutsites.csv\", \"r\")\n",
    "    pt = from_csv(fp)\n",
    "    fp.close()\n",
    "\n",
    "    print(pt)\n",
    "\n",
    "    cutsiteNum = int(input(\"Choose your cutsite (Leftmost number):\"))\n",
    "\n",
    "    print(\"\")\n",
    "    print(\"You've chosen: \")\n",
    "    print(pt.get_string(start = cutsiteNum-1, end = cutsiteNum))\n",
    "\n",
    "    counter = 0\n",
    "    for row in pt:\n",
    "        counter += 1\n",
    "        if counter == cutsiteNum:\n",
    "            row.border = False\n",
    "            row.header = False\n",
    "            cutsite = row.get_string(fields=[\"sequence\"]).strip()\n",
    "            cutName = row.get_string(fields=[\"name\"]).strip()\n",
    "            chromosome = int(row.get_string(fields=[\"chromosome\"]).strip())\n",
    "    if chromosome >= 10:\n",
    "        filename = \"Scer\" + str(chromosome) + \".fasta\" \n",
    "    else:\n",
    "        filename = \"Scer0\" + str(chromosome) + \".fasta\"\n",
    "\n",
    "    ChromosomeRec=SeqIO.read(filename, \"fasta\")\n",
    "    ChromosomeSeq = ChromosomeRec.seq\n",
    "\n",
    "    if ChromosomeSeq.find(cutsite) == -1:\n",
    "        ChromosomeSeq = ChromosomeSeq.reverse_complement()\n",
    "\n",
    "    startInd = ChromosomeSeq.find(cutsite)\n",
    "\n",
    "    UpHomology = SeqRecord(ChromosomeSeq[startInd - 1000:startInd])\n",
    "    DownHomology = SeqRecord(ChromosomeSeq[startInd + 30:startInd + 1030])\n",
    "    return UpHomology, DownHomology\n",
    "\n",
    "\n",
    "\n",
    "askUser()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.4.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
