{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "What's the name of the promoter? e.g. ADH1: TEF2\n",
      "\n",
      "Lupup AGTAGCATCTGTGGCAACAGATTC\n",
      "Rupup(TEF2ps) TATATATACATTAAACATTATTGGTAACAGGAATGGTTTAACTGCTAAGAGACA\n",
      "LTEF2ps(up) TAAACCATTCCTGTTACCAATAATGTTTAATGTATATATATATATATATATATGGGGCCG\n",
      "RTEF2ps(down) AAGTCTTGACGAAAATCTGCATGTTTAGTTAATTATAGTTCGTTGACCGTATATTCTAAA\n",
      "Ldowndown(TEF2ps) CGGTCAACGAACTATAATTAACTAAACATGCAGATTTTCGTCAAGACTTTGACC\n",
      "Rdowndown AGAAAATTTGTGCCCATTAACATCACC\n",
      "\n",
      "Your donor DNA cassette, has the following bp length and sequence:\n",
      "\n",
      "1379\n",
      "\n",
      "AGTAGCATCTGTGGCAACAGATTCTAGCCCTTTGAAGGATGAATGACTATTTTTTGTGCTAGTAGTGTAGAACACTTTATTTTGGGGAGCACTAGAGGAATATTGCGCCAACAATTGTGGGAGCGTCAAGAGGTCAGAAGCAGTCATTCGCCAATCTTTTGTTTTTCTCTGGTTCCCGTTCCCTACTGTTACTTTATTAGTATATAAGGGATCTGAATAAATATCAAGACGATGGAAAAAAAATAACAAGCAAAGGAACCTTTCTATATCCCAATTTTATTTCACATTAAAAACTATCGTCTTCCAACACTTTATATAACTGCTTGCAGCGGCCAGTAATTCGTCGCATCTCCATGACTCAAATTTTCCAGTGTCTCTTAGCAGTTAAACCATTCCTGTTACCAATAATGTTTAATGTATATATATATATATATATATGGGGCCGTATACTTACATATAGTAGATGTCAAGCGTAGGCGCTTCCCCTGCCGGCTGTGAGGGCGCCATAACCAAGGTATCTATAGACCGCCAATCAGCAAACTACCTCCGTACATTCATGTTGCACCCACACATTTATACACCCAGACCGCGACAAATTACCCATAAGGTTGTTTGTGACGGCGTCGTACAAGAGAACGTGGGAACTTTTTAGGCTCACCAAAAAAGAAAGAAAAAATACGAGTTGCTGACAGAAGCCTCAAGAAAAAAAAAATTCTTCTTCGACTATGCTGGAGGCAGAGATGATCGAGCCGGTAGTTAACTATATATAGCTAAATTGGTTCCATCACCTTCTTTTCTGGTGTCGCTCCTTCTAGTGCTATTTCTGGCTTTTCCTATTTTTTTTTTTCCATTTTTCTTTCTCTCTTTCTAATATATAAATTCTCTTGCATTTTCTATTTTTCTCTCTATCTATTCTACTTGTTTATTCCCTTCAAGGTTTTTTTTTAAGGAGTACTTGTTTTTAGAATATACGGTCAACGAACTATAATTAACTAAACATGCAGATTTTCGTCAAGACTTTGACCGGTAAAACCATAACATTGGAAGTTGAATCTTCCGATACCATCGACAACGTTAAGTCGAAAATTCAAGACAAGGAAGGTATCCCTCCAGATCAACAAAGATTGATCTTTGCCGGTAAGCAGCTAGAAGACGGTAGAACGCTGTCTGATTACAACATTCAGAAGGAGTCCACCTTACATCTTGTGCTAAGGCTAAGAGGTGGTTATCACGGATCCGGAGCTTGGCTGTTGCCCGTCTCACTGGTGAAAAGAAAAACCACCCTGGCGCCCAATACGAGTAAAGGAGAAGAACTTTTCACTGGAGTTGTCCCAATTCTTGTTGAATTAGATGGTGATGTTAATGGGCACAAATTTTCT\n",
      "\n",
      "You might want to copy this entire prompt and save it for your records.\n"
     ]
    }
   ],
   "source": [
    "#!/usr/bin/env python\n",
    "#title           : Promoters1.0001\n",
    "#description     : to design primers for promoters swap\n",
    "#author          : Leo d'Espaux <leodespaux@gmail.com>,  \n",
    "#date            : 20150423\n",
    "#version         : 3.0\n",
    "#notes           : includes options for \n",
    "#notes           : delGene replGene cutOligos\n",
    "#notes           : at some point there will be one interface\n",
    "#==============================================================================\n",
    "\n",
    "# import libraries we're using\n",
    "\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio import SeqIO\n",
    "from Bio.SeqUtils import MeltingTemp\n",
    "import copy\n",
    "from intermine.webservice import Service\n",
    "from pandas import *\n",
    "from pandas import DataFrame, read_csv\n",
    "import pandas as pd #this is how I usually import pandas\n",
    "import matplotlib.pyplot as plt\n",
    "import sys\n",
    "\n",
    "\n",
    "# define global variables\n",
    "HomologyLength = 1000\n",
    "PrimerMaxTm = 55 # As calculated by statluc, more like 65C by Thermo's Phusion program\n",
    "PrimerMaxLen = 60\n",
    "OverhangMaxFrac = 1\n",
    "\n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "def PromOligos():\n",
    "\n",
    "# This function asks user for a chromosomal locus, a region to be deleted, a suitable CRIPSR cutsite \n",
    "# and outputs oligos for cloning of a pL308 Cas9-gRNA vector, and ones for generating a donor DNA\n",
    "# to delete the unwanted chromosomal region. Primers Lup+Rdown produce a 1kb band if deletion was\n",
    "# successful. \n",
    "# part of yCRISPRv3 by leodespaux@gmail.com\n",
    "\n",
    "    \n",
    "    PromoterName=input(\"What's the name of the promoter? e.g. ADH1: \")\n",
    "    PromoterGeneRec=fetchGene(PromoterName)\n",
    "    PromoterRec=fetchNeighbor(PromoterGeneRec,\"upstream\",600)\n",
    "    PromoterRec.name=PromoterRec.name+\"ps\"\n",
    "    \n",
    "    \n",
    "    UpSeq=\"AGTAGCATCTGTGGCAACAGATTCTAGCCCTTTGAAGGATGAATGACTATTTTTTGTGCTAGTAGTGTAGAACACTTTATTTTGGGGAGCACTAGAGGAATATTGCGCCAACAATTGTGGGAGCGTCAAGAGGTCAGAAGCAGTCATTCGCCAATCTTTTGTTTTTCTCTGGTTCCCGTTCCCTACTGTTACTTTATTAGTATATAAGGGATCTGAATAAATATCAAGACGATGGAAAAAAAATAACAAGCAAAGGAACCTTTCTATATCCCAATTTTATTTCACATTAAAAACTATCGTCTTCCAACACTTTATATAACTGCTTGCAGCGGCCAGTAATTCGTCGCATCTCCATGACTCAAATTTTCCAGTGTCTCTTAGCAGTTAAACCATTCCTG\"\n",
    "    UpSeq=Seq(UpSeq)\n",
    "    UpRec=SeqRecord(UpSeq, name=\"up\")\n",
    "                   \n",
    "    DownSeq=\"ATGCAGATTTTCGTCAAGACTTTGACCGGTAAAACCATAACATTGGAAGTTGAATCTTCCGATACCATCGACAACGTTAAGTCGAAAATTCAAGACAAGGAAGGTATCCCTCCAGATCAACAAAGATTGATCTTTGCCGGTAAGCAGCTAGAAGACGGTAGAACGCTGTCTGATTACAACATTCAGAAGGAGTCCACCTTACATCTTGTGCTAAGGCTAAGAGGTGGTTATCACGGATCCGGAGCTTGGCTGTTGCCCGTCTCACTGGTGAAAAGAAAAACCACCCTGGCGCCCAATACGAGTAAAGGAGAAGAACTTTTCACTGGAGTTGTCCCAATTCTTGTTGAATTAGATGGTGATGTTAATGGGCACAAATTTTCT\"\n",
    "    DownSeq=Seq(DownSeq)\n",
    "    DownRec=SeqRecord(DownSeq, name=\"down\")\n",
    "    \n",
    "    fragments=[UpRec, PromoterRec, DownRec]\n",
    "    \n",
    "    stitch(fragments)\n",
    "    \n",
    "    \n",
    "def stitch(fragments):\n",
    "    #this function takes seq records and prints primers\n",
    "    \n",
    "    #let's make an empty sequence file\n",
    "    Nfrags=len(fragments)\n",
    "    donor=Seq(\"\")\n",
    "    index=[]\n",
    "    print(\"\")\n",
    "    for i in range (0, Nfrags):\n",
    "        donor=donor+fragments[i]\n",
    "    \n",
    "    for i in range (0, Nfrags):\n",
    "        if i==0:\n",
    "            print(\"Lup\"+ fragments[i].name + \" \" + getPrimer(donor))\n",
    "            print(\"Rup\"+ fragments[i].name + \"(\" + fragments[i+1].name + \") \" + overhangPrimer(fragments[i].reverse_complement(),fragments[i+1].reverse_complement()))\n",
    "        elif i==Nfrags-1:\n",
    "            print(\"Ldown\"+ fragments[i].name + \"(\" + fragments[i-1].name + \") \" + overhangPrimer(fragments[i],fragments[i-1]))\n",
    "            print(\"Rdown\"+ fragments[i].name + \" \" + getPrimer(donor.reverse_complement()))\n",
    "        else:\n",
    "            print(\"L\"+ fragments[i].name + \"(\" + fragments[i-1].name + \") \" + overhangPrimer(fragments[i],fragments[i-1]))\n",
    "            print(\"R\"+ fragments[i].name + \"(\" + fragments[i+1].name + \") \" + overhangPrimer(fragments[i].reverse_complement(),fragments[i+1].reverse_complement()))\n",
    "\n",
    "    print(\"\")\n",
    "    print(\"Your donor DNA cassette, has the following bp length and sequence:\")\n",
    "\n",
    "\n",
    "    print(\"\")\n",
    "    print(len(donor.seq))\n",
    "    print(\"\")\n",
    "\n",
    "    print(donor.seq)\n",
    "\n",
    "    print(\"\")\n",
    "    print(\"You might want to copy this entire prompt and save it for your records.\")\n",
    "    \n",
    "        \n",
    "def flipRecord(origRecord):\n",
    "    origRecord.seq=origRecord.seq.reverse_complement()\n",
    "    origRecord.name=\"r\"+origRecord.name\n",
    "    return origRecord\n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "def fetchGene(GeneName):\n",
    "    \n",
    "    service = Service(\"http://yeastmine.yeastgenome.org/yeastmine/service\")\n",
    "    template = service.get_template('Gene_GenomicDNA')\n",
    "\n",
    "    rows = template.rows(\n",
    "        E = {\"op\": \"LOOKUP\", \"value\": GeneName, \"extra_value\": \"S. cerevisiae\"}\n",
    "    )\n",
    "    \n",
    "    # this service seems to return multiple similar genes but we want the first one only, so count\n",
    "    # and it returns information about the gene you want\n",
    "    count=0\n",
    "    for row in rows:\n",
    "        \n",
    "        count=count+1\n",
    "        if count==1:\n",
    "            descr= row[\"description\"]\n",
    "            GeneSeq=Seq(row[\"sequence.residues\"])\n",
    "            GeneSysName=row[\"secondaryIdentifier\"]\n",
    "       \n",
    "    #let's create a record for the oldGene\n",
    "    GeneRecord = SeqRecord(GeneSeq, id=GeneSysName)\n",
    "    \n",
    "    #now let's add some more information to make it useful\n",
    "    GeneRecord.name=GeneName\n",
    "    GeneRecord.features=GeneSysName\n",
    "    GeneRecord.description=descr\n",
    "\n",
    "    return GeneRecord \n",
    "    \n",
    "\n",
    "def fetchNeighbor(NeighborRecord, direction, distance):\n",
    "\n",
    "    # let's load the appropriate chromosome file. The record of the gene we looked up\n",
    "    # contains in the \"features\" the systematic name, wherein the second letter\n",
    "    # corresponds to chromosome number, e.g., 1=A etc\n",
    "    if NeighborRecord.features[1]==\"A\":\n",
    "        ChromosomeRec=SeqIO.read(\"Scer01.fasta\", \"fasta\")\n",
    "    if NeighborRecord.features[1]==\"B\":\n",
    "        ChromosomeRec=SeqIO.read(\"Scer02.fasta\", \"fasta\")\n",
    "    if NeighborRecord.features[1]==\"C\":\n",
    "        ChromosomeRec=SeqIO.read(\"Scer03.fasta\", \"fasta\")\n",
    "    if NeighborRecord.features[1]==\"D\":\n",
    "        ChromosomeRec=SeqIO.read(\"Scer04.fasta\", \"fasta\")\n",
    "    if NeighborRecord.features[1]==\"E\":\n",
    "        ChromosomeRec=SeqIO.read(\"Scer05.fasta\", \"fasta\")\n",
    "    if NeighborRecord.features[1]==\"F\":\n",
    "        ChromosomeRec=SeqIO.read(\"Scer06.fasta\", \"fasta\")\n",
    "    if NeighborRecord.features[1]==\"G\":\n",
    "        ChromosomeRec=SeqIO.read(\"Scer07.fasta\", \"fasta\")\n",
    "    if NeighborRecord.features[1]==\"H\":\n",
    "        ChromosomeRec=SeqIO.read(\"Scer08.fasta\", \"fasta\")\n",
    "    if NeighborRecord.features[1]==\"I\":\n",
    "        ChromosomeRec=SeqIO.read(\"Scer09.fasta\", \"fasta\")\n",
    "    if NeighborRecord.features[1]==\"J\":\n",
    "        ChromosomeRec=SeqIO.read(\"Scer10.fasta\", \"fasta\")\n",
    "    if NeighborRecord.features[1]==\"K\":\n",
    "        ChromosomeRec=SeqIO.read(\"Scer11.fasta\", \"fasta\")\n",
    "    if NeighborRecord.features[1]==\"L\":\n",
    "        ChromosomeRec=SeqIO.read(\"Scer12.fasta\", \"fasta\")\n",
    "    if NeighborRecord.features[1]==\"M\":\n",
    "        ChromosomeRec=SeqIO.read(\"Scer13.fasta\", \"fasta\")\n",
    "    if NeighborRecord.features[1]==\"N\":\n",
    "        ChromosomeRec=SeqIO.read(\"Scer14.fasta\", \"fasta\")\n",
    "    if NeighborRecord.features[1]==\"O\":\n",
    "        ChromosomeRec=SeqIO.read(\"Scer15.fasta\", \"fasta\")\n",
    "    if NeighborRecord.features[1]==\"P\":\n",
    "        ChromosomeRec=SeqIO.read(\"Scer16.fasta\", \"fasta\") \n",
    "\n",
    "    \n",
    "    \n",
    "    # let's explicitely name the sequences from the seq record\n",
    "    NeighborSeq=NeighborRecord.seq\n",
    "    ChromosomeSeq=ChromosomeRec.seq\n",
    "    \n",
    "    # flip the sequence to orient with respect to the old gene\n",
    "    if ChromosomeSeq.find(NeighborSeq)==-1:\n",
    "        ChromosomeSeq=ChromosomeSeq.reverse_complement()\n",
    "\n",
    "    StartIndex=ChromosomeSeq.find(NeighborSeq)\n",
    "    EndIndex=StartIndex+len(NeighborSeq)\n",
    "    \n",
    "    if direction==\"upstream\":\n",
    "        DesiredSeq=ChromosomeSeq[StartIndex-distance:StartIndex]\n",
    "    if direction==\"downstream\":\n",
    "        DesiredSeq=ChromosomeSeq[EndIndex:EndIndex+distance]\n",
    "\n",
    "    \n",
    "    \n",
    "    \n",
    "    NeighborRec = SeqRecord(DesiredSeq, name=NeighborRecord.name)\n",
    "    \n",
    "    return NeighborRec\n",
    "\n",
    "    \n",
    "def getPrimer(currRecord):\n",
    "    \n",
    "\n",
    "    mp = 0\n",
    "    length = 0\n",
    "    primer = Seq(\"\")\n",
    "\n",
    "    seq=currRecord.seq\n",
    "    \n",
    "    while mp <= PrimerMaxTm and length <= PrimerMaxLen:\n",
    "        primer = primer + seq[length]\n",
    "        mp = MeltingTemp.Tm_staluc(primer)\n",
    "        length += 1\n",
    "\n",
    "    return primer           \n",
    "        \n",
    "def overhangPrimer(currRecord,prevSeq):\n",
    "    #let's get the template-binding primer first\n",
    "    primer=getPrimer(currRecord)\n",
    "    \n",
    "    \n",
    "    #OK let's work on the overhang\n",
    "    maxOhLen=PrimerMaxLen-len(primer)    \n",
    "    maxFrac=1\n",
    "    \n",
    "    #let's decide on a max overhang length\n",
    "    if round(len(primer)*(OverhangMaxFrac+1)) < 60:\n",
    "             maxOhLen=round(len(primer)*OverhangMaxFrac)\n",
    "    \n",
    "    #the index must be an integer!!!\n",
    "    maxOhLen=int(maxOhLen)\n",
    "    ohprimer=prevSeq.seq[-maxOhLen:]+primer #we add the .seq so that it returns a string\n",
    "    \n",
    "    return ohprimer      \n",
    "        \n",
    "def standardCassette():\n",
    "    \n",
    "    #first, the promoter\n",
    "    print(\"I'm going to build a standard cassette in which promoter is 600nt, terminator 250nt.\")\n",
    "    print(\"\")\n",
    "    print(\"Which PROMOTER do you want to use, e.g., TDH3\")\n",
    "    \n",
    "    PromoterName=input(\"Your answer: \")\n",
    "    PromoterGeneRec=fetchGene(PromoterName)\n",
    "    PromoterRec=fetchNeighbor(PromoterGeneRec,\"upstream\",600)\n",
    "    PromoterRec.name=PromoterRec.name+\"ps\"\n",
    "    \n",
    "    \n",
    "    #second, the terminator\n",
    "    print(\"Which TERMINATOR do you want to use, e.g., ADH1\")\n",
    "    TerminatorName = input('Your answer: ')\n",
    "    TerminatorGeneRec=fetchGene(TerminatorName)\n",
    "    TerminatorRec=fetchNeighbor(TerminatorGeneRec,\"downstream\",250)\n",
    "    TerminatorRec.name=TerminatorRec.name+\"ts\"\n",
    "    \n",
    "    \n",
    "    #and last, the gene\n",
    "    print(\"What do you want to call the CDS you're inserting?\")\n",
    "    orfName = input(\"Your answer: \")\n",
    "    \n",
    "    print(\"What's the sequence\")\n",
    "    orfSeq=input(\"Your answer: \")\n",
    "    \n",
    "    orfRecord=SeqRecord(Seq(orfSeq), name=orfName)\n",
    "    \n",
    "    insertRec=[PromoterRec,orfRecord,TerminatorRec]\n",
    "    return PromoterRec, orfRecord, TerminatorRec\n",
    "          \n",
    "def buildCassette():\n",
    "    \n",
    "    #first, the promoter\n",
    "    print(\"I'm going to build a standard cassette in which promoter is 600nt, terminator 250nt.\") \n",
    "    print(\"First, which PROMOTER do you want to use, e.g., TDH3\")\n",
    "    \n",
    "    PromoterName=input(\"Your answer: \")\n",
    "    PromoterGeneRec=fetchGene(PromoterName)\n",
    "    PromoterRec=fetchNeighbor(PromoterGeneRec,\"upstream\",600)\n",
    "    PromoterRec.name=PromoterRec.name+\"ps\"\n",
    "    \n",
    "    \n",
    "    #second, the terminator\n",
    "    print(\"Which TERMINATOR do you want to use, e.g., ADH1\")\n",
    "    TerminatorName = input('Your answer: ')\n",
    "    TerminatorGeneRec=fetchGene(TerminatorName)\n",
    "    TerminatorRec=fetchNeighbor(TerminatorGeneRec,\"downstream\",250)\n",
    "    TerminatorRec.name=TerminatorRec.name+\"ts\"\n",
    "    \n",
    "    #and last, the gene\n",
    "    print(\"What is the name of your gene, e.g., KlGapDH\")\n",
    "    orfName = input(\"Your answer: \")\n",
    "    \n",
    "    print(\"What's the sequence\")\n",
    "    orfSeq=input(\"Your answer: \")\n",
    "    \n",
    "    orfRecord=SeqRecord(Seq(orfSeq), name=orfName)\n",
    "    \n",
    "    insertRec=[PromoterRec,orfRecord,TerminatorRec]\n",
    "    return PromoterRec, orfRecord, TerminatorRec\n",
    "    \n",
    "\n",
    "# now to run, type the program and ^Enter or play. E.g. remove hash below to run delGene()\n",
    "\n",
    "PromOligos()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.4.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
