{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Python version 3.4.3 |Anaconda 2.3.0 (x86_64)| (default, Mar  6 2015, 12:07:41) \n",
      "[GCC 4.2.1 (Apple Inc. build 5577)]\n",
      "IMPORTANT: IF THE ABOVE SAYS PYTHON 3.X DO NOT USE QUOTES IN RESPONSES.\n",
      " \n",
      "Hi baby, welcome to SynBioCAD. I'm here for you.\n",
      "Right now, we assume you're working in S. cerevisiae.\n",
      " \n",
      "What do you want to do today?\n",
      "\"1\" to insert DNA into a characterized locus\n",
      "\"2\" to edit an existing gene, e.g., delete or replace\n",
      "\"3\" to build a custom cassette\n",
      " \n",
      "Your answer: 1\n",
      " \n",
      "      chrom. loc.  exp. lev.   name              sequence\n",
      "208a       Scer02        1.0   208a  GTCCGCTAAACAAAAGATCT\n",
      "1014a      Scer10        1.2  1014a  TTATGTGCGTATTGCTTTCA\n",
      "1114a      Scer11        1.1  1114a  CTTGTGAAACAAATAATTGG\n",
      "607c       Scer06        1.1   607c  CTATTTTTGCTTTCTGCACA\n",
      "308a       Scer03        1.4   308a  TAGGATACAAGTGTCTGTGC\n",
      "1021b      Scer10        1.0  1021b  CCTCTGTGTGGTGGTAATTG\n",
      "720a       Scer07        1.0   720a  CAACAATTGTTACAATAGTA\n",
      "\n",
      "Which cut site do you want, e.g., 208a?\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m/Users/leodespaux/anaconda/lib/python3.4/site-packages/IPython/kernel/zmq/kernelbase.py\u001b[0m in \u001b[0;36m_input_request\u001b[0;34m(self, prompt, ident, parent, password)\u001b[0m\n\u001b[1;32m    675\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 676\u001b[0;31m                 \u001b[0mident\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mreply\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msession\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrecv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstdin_socket\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    677\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mException\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/leodespaux/anaconda/lib/python3.4/site-packages/IPython/kernel/zmq/session.py\u001b[0m in \u001b[0;36mrecv\u001b[0;34m(self, socket, mode, content, copy)\u001b[0m\n\u001b[1;32m    718\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 719\u001b[0;31m             \u001b[0mmsg_list\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msocket\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrecv_multipart\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmode\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcopy\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mcopy\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    720\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0mzmq\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mZMQError\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/leodespaux/anaconda/lib/python3.4/site-packages/zmq/sugar/socket.py\u001b[0m in \u001b[0;36mrecv_multipart\u001b[0;34m(self, flags, copy, track)\u001b[0m\n\u001b[1;32m    357\u001b[0m         \"\"\"\n\u001b[0;32m--> 358\u001b[0;31m         \u001b[0mparts\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrecv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mflags\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcopy\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mcopy\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtrack\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mtrack\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    359\u001b[0m         \u001b[0;31m# have first part already, only loop while more to receive\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32mzmq/backend/cython/socket.pyx\u001b[0m in \u001b[0;36mzmq.backend.cython.socket.Socket.recv (zmq/backend/cython/socket.c:6507)\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mzmq/backend/cython/socket.pyx\u001b[0m in \u001b[0;36mzmq.backend.cython.socket.Socket.recv (zmq/backend/cython/socket.c:6307)\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mzmq/backend/cython/socket.pyx\u001b[0m in \u001b[0;36mzmq.backend.cython.socket._recv_copy (zmq/backend/cython/socket.c:1877)\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32m/Users/leodespaux/anaconda/lib/python3.4/site-packages/zmq/backend/cython/checkrc.pxd\u001b[0m in \u001b[0;36mzmq.backend.cython.checkrc._check_rc (zmq/backend/cython/socket.c:6758)\u001b[0;34m()\u001b[0m\n\u001b[1;32m     11\u001b[0m     \u001b[0mcdef\u001b[0m \u001b[0mint\u001b[0m \u001b[0merrno\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mzmq_errno\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 12\u001b[0;31m     \u001b[0mPyErr_CheckSignals\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     13\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mrc\u001b[0m \u001b[0;34m<\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: ",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-1-7cf8b8a94a1a>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m    578\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    579\u001b[0m \u001b[0;31m#print (buildCassette())\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 580\u001b[0;31m \u001b[0maskUser\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    581\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-1-7cf8b8a94a1a>\u001b[0m in \u001b[0;36maskUser\u001b[0;34m()\u001b[0m\n\u001b[1;32m     51\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mAction\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m\"1\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     52\u001b[0m         \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\" \"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 53\u001b[0;31m         \u001b[0meditEmpty\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     54\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     55\u001b[0m     \u001b[0;32melif\u001b[0m \u001b[0mAction\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m\"2\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-1-7cf8b8a94a1a>\u001b[0m in \u001b[0;36meditEmpty\u001b[0;34m()\u001b[0m\n\u001b[1;32m     83\u001b[0m     \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     84\u001b[0m     \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Which cut site do you want, e.g., 208a?\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 85\u001b[0;31m     \u001b[0mcutname\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0minput\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Your answer: \"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     86\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     87\u001b[0m     \u001b[0mlocation\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mcutFrame\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mcutname\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m'chrom. loc.'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0;34m\".fasta\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/leodespaux/anaconda/lib/python3.4/site-packages/IPython/kernel/zmq/kernelbase.py\u001b[0m in \u001b[0;36mraw_input\u001b[0;34m(self, prompt)\u001b[0m\n\u001b[1;32m    649\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_parent_ident\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    650\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_parent_header\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 651\u001b[0;31m             \u001b[0mpassword\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    652\u001b[0m         )\n\u001b[1;32m    653\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/leodespaux/anaconda/lib/python3.4/site-packages/IPython/kernel/zmq/kernelbase.py\u001b[0m in \u001b[0;36m_input_request\u001b[0;34m(self, prompt, ident, parent, password)\u001b[0m\n\u001b[1;32m    679\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mKeyboardInterrupt\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    680\u001b[0m                 \u001b[0;31m# re-raise KeyboardInterrupt, to truncate traceback\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 681\u001b[0;31m                 \u001b[0;32mraise\u001b[0m \u001b[0mKeyboardInterrupt\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    682\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    683\u001b[0m                 \u001b[0;32mbreak\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "#!/usr/bin/env python\n",
    "#title           : SynBioCAD1.0 beta\n",
    "#description     : defines several methods to generate oligos for CRISPR\n",
    "#author          : Leo d'Espaux <leodespaux@gmail.com>\n",
    "#date            : 9 June 2015\n",
    "#==============================================================================\n",
    "\n",
    "\n",
    "# import libraries we're using\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio import SeqIO\n",
    "from Bio.SeqUtils import MeltingTemp\n",
    "import copy\n",
    "\n",
    "from intermine.webservice import Service\n",
    "\n",
    "from pandas import *\n",
    "from pandas import DataFrame, read_csv\n",
    "import pandas as pd #this is how I usually import pandas\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "import sys\n",
    "\n",
    "# define global variables\n",
    "HomologyLength = 1000\n",
    "PrimerMaxTm = 55\n",
    "PrimerMaxLen = 60\n",
    "OverhangMaxFrac = 1\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "def askUser():\n",
    "    print('Python version ' + sys.version)\n",
    "    print(\"IMPORTANT: IF THE ABOVE SAYS PYTHON 3.X DO NOT USE QUOTES IN RESPONSES.\")\n",
    "    #print('Pandas version ' + pd.__version__)\n",
    "    print(\" \")\n",
    "    print(\"Hi baby, welcome to SynBioCAD. I'm here for you.\")\n",
    "    print(\"Right now, we assume you're working in S. cerevisiae.\")\n",
    "    print(\" \")\n",
    "    print(\"What do you want to do today?\")\n",
    "    print(\"\\\"1\\\" to insert DNA into a characterized locus\") \n",
    "    print(\"\\\"2\\\" to edit an existing gene, e.g., delete or replace\")\n",
    "    print(\"\\\"3\\\" to build a custom cassette\")\n",
    "    print(\" \")\n",
    "\n",
    "    Action= input(\"Your answer: \")\n",
    "    if Action == \"1\":\n",
    "        print(\" \")\n",
    "        editEmpty()\n",
    "        \n",
    "    elif Action == \"2\":\n",
    "        print(\" \")\n",
    "        editExisting()\n",
    "\n",
    "    elif Action == \"3\":\n",
    "        print(\" \")\n",
    "        buildCustom()\n",
    "              \n",
    "def editEmpty():\n",
    "    \n",
    "    labels=['208a', '1014a', '1114a', '607c', '308a', '1021b', '720a']\n",
    "    ChrLetters=[\"Scer02\", \"Scer10\", \"Scer11\", \"Scer06\", \"Scer03\", \"Scer10\", \"Scer07\"]\n",
    "    ExpValues=[1.0, 1.2, 1.1, 1.1, 1.4, 1.0, 1.0]\n",
    "    cutSeqs=[\"GTCCGCTAAACAAAAGATCT\", \"TTATGTGCGTATTGCTTTCA\", \"CTTGTGAAACAAATAATTGG\", \"CTATTTTTGCTTTCTGCACA\", \"TAGGATACAAGTGTCTGTGC\", \"CCTCTGTGTGGTGGTAATTG\", \"CAACAATTGTTACAATAGTA\"]\n",
    "\n",
    "    \n",
    "    #We save information as an array\n",
    "    cutArray={'name' : Series(labels, index=labels),\n",
    "                'exp. lev.' : Series(ExpValues, index=labels),\n",
    "                'chrom. loc.' : Series(ChrLetters, index=labels),\n",
    "                'sequence' : Series(cutSeqs, index=labels)\n",
    "             }\n",
    "\n",
    "    \n",
    "    cutFrame= DataFrame(cutArray)\n",
    "    \n",
    "    print(cutFrame)\n",
    "    \n",
    "    print(\"\")\n",
    "    print(\"Which cut site do you want, e.g., 208a?\")\n",
    "    cutname=input(\"Your answer: \")\n",
    "\n",
    "    location=cutFrame.loc[cutname,'chrom. loc.']+\".fasta\"\n",
    "    cutSequence=cutFrame.loc[cutname,'sequence']\n",
    "    \n",
    "    \n",
    "    # we use that location to get the chromosome sequence\n",
    "    ChromosomeSeq=SeqIO.read(location, \"fasta\").seq\n",
    "    \n",
    "    if ChromosomeSeq.find(cutSequence)==-1:\n",
    "        ChromosomeSeq=ChromosomeSeq.reverse_complement()\n",
    "\n",
    "    StartIndex=ChromosomeSeq.find(cutSequence)\n",
    "    EndIndex=StartIndex+34\n",
    "    \n",
    "    UpSeq=ChromosomeSeq[StartIndex-HomologyLength:StartIndex]\n",
    "    DownSeq=ChromosomeSeq[EndIndex:EndIndex+HomologyLength]\n",
    "        \n",
    "    UpHomRec = SeqRecord(UpSeq, id=cutname)\n",
    "    DownHomRec = SeqRecord(DownSeq, id=cutname)\n",
    "    \n",
    "    print(\"\")\n",
    "    print(\"Found your homology regions\")\n",
    "    print(\"\")\n",
    "\n",
    "              \n",
    "    print(\"What do you want to do to \" + cutname + \"?\")\n",
    "    print(\"\\\"1\\\" to insert a pre-built custom cassette\")\n",
    "    print(\"\\\"2\\\" to construct a cassette using standard promoter and terminator names, and a custom ORF\")\n",
    "    print(\" \")\n",
    "\n",
    "    typeEdit=input(\"Your answer: \")\n",
    "    if typeEdit==\"1\":\n",
    "    \n",
    "        orfName=input(\"What's the name of your custom gene or cassette?\")\n",
    "        orfSeq=Seq(input(\"What's the sequence?\"))\n",
    "        \n",
    "        orfRecord=SeqRecord(orfSeq,id=orfName)\n",
    "        \n",
    "        fragments=[UpHomRec, orfRecord, DownHomRec]\n",
    "\n",
    "    \n",
    "    elif typeEdit == \"2\":\n",
    "        \n",
    "        PromoterRec, orfRecord, TerminatorRec = buildCassette()\n",
    "        \n",
    "        fragments=[UpHomRec, PromoterRec, orfRecord, TerminatorRec, DownHomRec]\n",
    "\n",
    "    \n",
    "    stitch(fragments)\n",
    "    \n",
    "def editExisting():\n",
    "    print(\" \")\n",
    "    print(\"Which locus do you want to edit? Tell me a common name, e.g., \\\"OAF1\\\": \")\n",
    "    print(\"I'm smart and pretty and I can fetch it for you.\")\n",
    "    \n",
    "    GeneName= input(\"Your answer: \")\n",
    "    \n",
    "    print(\"\")\n",
    "    print(\"Give me a second to find it for you...\")\n",
    "\n",
    "        \n",
    "    OrigGeneRecord=fetchGene(GeneName)\n",
    "    #note that this returns a seqrecord\n",
    "    \n",
    "    print(\"\")\n",
    "    print(\"I think you want \"+OrigGeneRecord.features+\":\")\n",
    "    print(OrigGeneRecord.description)\n",
    "\n",
    "\n",
    "    # We make seqrecords since that's what we carry through later in the program\n",
    "    UpHomRec = fetchNeighbor(OrigGeneRecord, \"upstream\", HomologyLength )\n",
    "    DownHomRec = fetchNeighbor(OrigGeneRecord, \"downstream\", HomologyLength )\n",
    "    \n",
    "    #print(\"OK I found them.\")\n",
    "\n",
    "    #print(\" \")\n",
    "    #print(\"UpstreamHomology reghhion is\")\n",
    "    #print(UpHomRec.seq)\n",
    "    \n",
    "    #print(\" \")\n",
    "    #print(\"DownstreamHomology reghhion is\")\n",
    "    #print(DownHomRec.seq)\n",
    "    \n",
    "    print(\" \")\n",
    "    print(\"What do you want to do to this gene?\")\n",
    "    print(\"\\\"1\\\" delete the protein coding DNA sequence (CDS)\")\n",
    "    print(\"\\\"2\\\" replace the CDS with another CDS, or other pre-made fragment\")\n",
    "    print(\"\\\"3\\\" replace the CDS with a standard cassette I will build for you\")\n",
    "    print(\"\\\"4\\\" replace the CDS with a custom cassette\")\n",
    "    print(\"\\\"5\\\" replace a specified region near your target gene\")\n",
    "    \n",
    "    print(\" \")\n",
    "    Action=input(\"Your answer: \")\n",
    "                        \n",
    "    #note that in all the below, we want to have fragments be records\n",
    "    if Action==\"1\":\n",
    "        fragments=[UpHomRec,DownHomRec]\n",
    "    \n",
    "    if Action==\"2\":\n",
    "        print(\" \")\n",
    "        NewGeneName=input(\"What's the name of the gene you're inserting?\")\n",
    "        NewGeneSeq=Seq(input(\"What's the sequence of your new gene?\"))\n",
    "        InsertRec = SeqRecord(NewGeneSeq, id=NewGeneName)\n",
    "        fragments=[UpHomRec, InsertRec, DownHomRec]\n",
    "\n",
    "    if Action==\"3\":\n",
    "        print(\" \")\n",
    "        print(\"I got you baby. I don't want you to strain your eyes looking up stuff.\")\n",
    "        PromoterRec, orfRecord, TerminatorRec = standardCassette()\n",
    "        fragments=[UpHomRec, PromoterRec, orfRecord, TerminatorRec, DownHomRec] #we need to finish buildcassette to add InsertRec here\n",
    "\n",
    "    if Action==\"4\":\n",
    "        print(\"How many pieces (other than homology fragments) are you stitching together.\")\n",
    "        Npieces=input(\"Your answer: \")\n",
    "        output = standardCassette()\n",
    "\n",
    "        fragments=[UpHomRec, DownHomRec] #we need to finish buildcassette to add InsertRec here\n",
    "\n",
    "        \n",
    "    stitch(fragments)\n",
    "\n",
    "def buildCustom():\n",
    "    print(\"How do you want your cassettes?\")\n",
    "    print(\"\\\"1\\\" to build a cassette with N pieces and no variants.\")\n",
    "    print(\"\\\"2\\\" to build a cassette with N pieces and one variant site.\")\n",
    "\n",
    "    answer = input(\"Your answer:\")\n",
    "\n",
    "    if answer == \"1\":\n",
    "        N = int(input(\"How many pieces (N) in your custom cassette:\"))\n",
    "        fragments = variableCassette(N)[0]\n",
    "        stitch(fragments)\n",
    "\n",
    "\n",
    "    elif answer == \"2\":\n",
    "        N = int(input(\"How many pieces (N) are in your custom cassette: \"))\n",
    "        toVary = int(input(\"Which piece would you like to vary: \"))\n",
    "        variants = int(input(\"How many variants would you like: \"))\n",
    "        fragsList = variableCassette(N, toVary, variants)\n",
    "        # Make the entire thing first. Then just primers for the variant.\n",
    "        print(\"\")\n",
    "        print(\"First we'll get primers for the entire cassette\")\n",
    "        stitch(fragsList[0])\n",
    "        print(\"-------------------------------------------------\")\n",
    "        print(\"Now we'll get primers just for the variants\")\n",
    "\n",
    "        for n in range(N-1):\n",
    "            print(\"For variant \" + str(n) + \":\")\n",
    "            frags = fragsList[n + 1][toVary - 2: toVary + 1]\n",
    "            stitch(frags)\n",
    "    #------------------------------ FETCH FUNCTIONS -------------------------------------\n",
    "\n",
    "def fetchGene(GeneName):\n",
    "    \n",
    "    service = Service(\"http://yeastmine.yeastgenome.org/yeastmine/service\")\n",
    "    template = service.get_template('Gene_GenomicDNA')\n",
    "\n",
    "    rows = template.rows(\n",
    "        E = {\"op\": \"LOOKUP\", \"value\": GeneName, \"extra_value\": \"S. cerevisiae\"}\n",
    "    )\n",
    "    \n",
    "    # this service seems to return multiple similar genes but we want the first one only, so count\n",
    "    # and it returns information about the gene you want\n",
    "    count=0\n",
    "    for row in rows:\n",
    "        \n",
    "        count=count+1\n",
    "        if count==1:\n",
    "            descr= row[\"description\"]\n",
    "            GeneSeq=Seq(row[\"sequence.residues\"])\n",
    "            GeneSysName=row[\"secondaryIdentifier\"]\n",
    "            #print(\" \")\n",
    "            #print(\"I think you want...... \"+row[\"secondaryIdentifier\"])\n",
    "            #print(row[\"description\"])\n",
    "            #print(\" \")\n",
    "            #print(row[\"sequence.residues\"])\n",
    "            #print(\" \")\n",
    "            #print(\"Good choice! I have a feeling you're going to get lucky with this one.\")\n",
    "            #print(\" \")\n",
    "            #print(\"Give me a second to put some of my ducks in a circle...\")\n",
    "       \n",
    "\n",
    "            \n",
    "    #let's create a record for the oldGene\n",
    "    GeneRecord = SeqRecord(GeneSeq, id=GeneSysName)\n",
    "    \n",
    "    #now let's add some more information to make it useful\n",
    "    GeneRecord.name=GeneName\n",
    "    GeneRecord.features=GeneSysName\n",
    "    GeneRecord.description=descr\n",
    "\n",
    "    return GeneRecord \n",
    "       \n",
    "def fetchNeighbor(NeighborRecord, direction, distance):\n",
    "\n",
    "\n",
    "    # let's load the appropriate chromosome file. The record of the gene we looked up\n",
    "    # contains in the \"features\" the systematic name, wherein the second letter\n",
    "    # corresponds to chromosome number, e.g., 1=A etc\n",
    "    if NeighborRecord.features[1]==\"A\":\n",
    "        ChromosomeRec=SeqIO.read(\"Scer01.fasta\", \"fasta\")\n",
    "    if NeighborRecord.features[1]==\"B\":\n",
    "        ChromosomeRec=SeqIO.read(\"Scer02.fasta\", \"fasta\")\n",
    "    if NeighborRecord.features[1]==\"C\":\n",
    "        ChromosomeRec=SeqIO.read(\"Scer03.fasta\", \"fasta\")\n",
    "    if NeighborRecord.features[1]==\"D\":\n",
    "        ChromosomeRec=SeqIO.read(\"Scer04.fasta\", \"fasta\")\n",
    "    if NeighborRecord.features[1]==\"E\":\n",
    "        ChromosomeRec=SeqIO.read(\"Scer05.fasta\", \"fasta\")\n",
    "    if NeighborRecord.features[1]==\"F\":\n",
    "        ChromosomeRec=SeqIO.read(\"Scer06.fasta\", \"fasta\")\n",
    "    if NeighborRecord.features[1]==\"G\":\n",
    "        ChromosomeRec=SeqIO.read(\"Scer07.fasta\", \"fasta\")\n",
    "    if NeighborRecord.features[1]==\"H\":\n",
    "        ChromosomeRec=SeqIO.read(\"Scer08.fasta\", \"fasta\")\n",
    "    if NeighborRecord.features[1]==\"I\":\n",
    "        ChromosomeRec=SeqIO.read(\"Scer09.fasta\", \"fasta\")\n",
    "    if NeighborRecord.features[1]==\"J\":\n",
    "        ChromosomeRec=SeqIO.read(\"Scer10.fasta\", \"fasta\")\n",
    "    if NeighborRecord.features[1]==\"K\":\n",
    "        ChromosomeRec=SeqIO.read(\"Scer11.fasta\", \"fasta\")\n",
    "    if NeighborRecord.features[1]==\"L\":\n",
    "        ChromosomeRec=SeqIO.read(\"Scer12.fasta\", \"fasta\")\n",
    "    if NeighborRecord.features[1]==\"M\":\n",
    "        ChromosomeRec=SeqIO.read(\"Scer13.fasta\", \"fasta\")\n",
    "    if NeighborRecord.features[1]==\"N\":\n",
    "        ChromosomeRec=SeqIO.read(\"Scer14.fasta\", \"fasta\")\n",
    "    if NeighborRecord.features[1]==\"O\":\n",
    "        ChromosomeRec=SeqIO.read(\"Scer15.fasta\", \"fasta\")\n",
    "    if NeighborRecord.features[1]==\"P\":\n",
    "        ChromosomeRec=SeqIO.read(\"Scer16.fasta\", \"fasta\") \n",
    "\n",
    "    \n",
    "    \n",
    "    # let's explicitely name the sequences from the seq record\n",
    "    NeighborSeq=NeighborRecord.seq\n",
    "    ChromosomeSeq=ChromosomeRec.seq\n",
    "    \n",
    "    # flip the sequence to orient with respect to the old gene\n",
    "    if ChromosomeSeq.find(NeighborSeq)==-1:\n",
    "        ChromosomeSeq=ChromosomeSeq.reverse_complement()\n",
    "\n",
    "    StartIndex=ChromosomeSeq.find(NeighborSeq)\n",
    "    EndIndex=StartIndex+len(NeighborSeq)\n",
    "    \n",
    "    if direction==\"upstream\":\n",
    "        DesiredSeq=ChromosomeSeq[StartIndex-distance:StartIndex]\n",
    "    if direction==\"downstream\":\n",
    "        DesiredSeq=ChromosomeSeq[EndIndex:EndIndex+distance]\n",
    "\n",
    "    \n",
    "    \n",
    "    \n",
    "    NeighborRec = SeqRecord(DesiredSeq, id=NeighborRecord.name)\n",
    "    \n",
    "    return NeighborRec\n",
    "\n",
    "    #print(NeighborRec)\n",
    "\n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    #------------------------------------ CONSTRUCTING STUFF --------------------------------------\n",
    "\n",
    "def getPrimer(currRecord):\n",
    "    \n",
    "\n",
    "    mp = 0\n",
    "    length = 0\n",
    "    primer = Seq(\"\")\n",
    "\n",
    "    seq=currRecord.seq\n",
    "    \n",
    "    while mp <= PrimerMaxTm and length <= PrimerMaxLen:\n",
    "        primer = primer + seq[length]\n",
    "        mp = MeltingTemp.Tm_staluc(primer)\n",
    "        length += 1\n",
    "\n",
    "    return primer           \n",
    "        \n",
    "def overhangPrimer(currRecord,prevSeq):\n",
    "    #let's get the template-binding primer first\n",
    "    primer=getPrimer(currRecord)\n",
    "    \n",
    "    \n",
    "    #OK let's work on the overhang\n",
    "    maxOhLen=PrimerMaxLen-len(primer)    \n",
    "    maxFrac=1\n",
    "    \n",
    "    #let's decide on a max overhang length\n",
    "    if round(len(primer)*(OverhangMaxFrac+1)) < 60:\n",
    "             maxOhLen=round(len(primer)*OverhangMaxFrac)\n",
    "    \n",
    "    #the index must be an integer!!!\n",
    "    maxOhLen=int(maxOhLen)\n",
    "    ohprimer=prevSeq.seq[-maxOhLen:]+primer #we add the .seq so that it returns a string\n",
    "    \n",
    "    return ohprimer      \n",
    "        \n",
    "def standardCassette():\n",
    "    \n",
    "    #first, the promoter\n",
    "    print(\"I'm going to build a standard cassette in which promoter is 600nt, terminator 250nt.\") \n",
    "    print(\"First, which PROMOTER do you want to use, e.g., TDH3\")\n",
    "    \n",
    "    PromoterName=input(\"Your answer: \")\n",
    "    PromoterGeneRec=fetchGene(PromoterName)\n",
    "    PromoterRec=fetchNeighbor(PromoterGeneRec,\"upstream\",600)\n",
    "    PromoterRec.id=PromoterRec.id+\"ps\"\n",
    "    \n",
    "    \n",
    "    #second, the terminator\n",
    "    print(\"Which TERMINATOR do you want to use, e.g., ADH1\")\n",
    "    TerminatorName = input('Your answer: ')\n",
    "    TerminatorGeneRec=fetchGene(TerminatorName)\n",
    "    TerminatorRec=fetchNeighbor(TerminatorGeneRec,\"downstream\",250)\n",
    "    TerminatorRec.id=TerminatorRec.id+\"ts\"\n",
    "    \n",
    "    \n",
    "    #and last, the gene\n",
    "    print(\"What is the name of your gene, e.g., KlGapDH\")\n",
    "    orfName = input(\"Your answer: \")\n",
    "    \n",
    "    print(\"What's the sequence\")\n",
    "    orfSeq=input(\"Your answer: \")\n",
    "    \n",
    "    orfRecord=SeqRecord(Seq(orfSeq), id=orfName)\n",
    "    \n",
    "    insertRec=[PromoterRec,orfRecord,TerminatorRec]\n",
    "    return PromoterRec, orfRecord, TerminatorRec\n",
    "          \n",
    "def buildCassette():\n",
    "    \n",
    "    #first, the promoter\n",
    "    print(\"I'm going to build a standard cassette in which promoter is 600nt, terminator 250nt.\") \n",
    "    print(\"First, which PROMOTER do you want to use, e.g., TDH3\")\n",
    "    \n",
    "    PromoterName=input(\"Your answer: \")\n",
    "    PromoterGeneRec=fetchGene(PromoterName)\n",
    "    PromoterRec=fetchNeighbor(PromoterGeneRec,\"upstream\",600)\n",
    "    PromoterRec.id=PromoterRec.id+\"ps\"\n",
    "    \n",
    "    \n",
    "    #second, the terminator\n",
    "    print(\"Which TERMINATOR do you want to use, e.g., ADH1\")\n",
    "    TerminatorName = input('Your answer: ')\n",
    "    TerminatorGeneRec=fetchGene(TerminatorName)\n",
    "    TerminatorRec=fetchNeighbor(TerminatorGeneRec,\"downstream\",250)\n",
    "    TerminatorRec.id=TerminatorRec.id+\"ts\"\n",
    "    \n",
    "    #and last, the gene\n",
    "    print(\"What is the name of your gene, e.g., KlGapDH\")\n",
    "    orfName = input(\"Your answer: \")\n",
    "    \n",
    "    print(\"What's the sequence\")\n",
    "    orfSeq=input(\"Your answer: \")\n",
    "    \n",
    "    orfRecord=SeqRecord(Seq(orfSeq), id=orfName)\n",
    "    \n",
    "    insertRec=[PromoterRec,orfRecord,TerminatorRec]\n",
    "    return PromoterRec, orfRecord, TerminatorRec\n",
    "    \n",
    "def variableCassette(N, toVary = 0, variants = 0):\n",
    "    print(\"\")\n",
    "    print(\"Let's start building.\")\n",
    "    print(\"\")\n",
    "    \n",
    "    # Store both name and sequence in a SeqRecord\n",
    "    # Append them to a list\n",
    "    # Return list as fragments to be stitched\n",
    "\n",
    "    records = []\n",
    "    for n in range(N):\n",
    "        name = input(\"What is the name of sequence \" + str(n+1) +\":\")\n",
    "        sequence = input(\"What is the sequence of this fragment:\")\n",
    "        print(\"\")\n",
    "        Rec = SeqRecord(Seq(sequence), id = name)\n",
    "        Rec.name = name\n",
    "        records.append(Rec)\n",
    "\n",
    "    variantRecords = []\n",
    "    variantRecords.append(records)\n",
    "    # This only happens if there are variants.\n",
    "    if variants > 0:\n",
    "        print(\"Time to make those variants you wanted.\")\n",
    "        for n in range(variants-1):\n",
    "            name = input(\"What is the name of variant \" + str(n+1) + \":\")\n",
    "            sequence = input(\"What is the sequence of this variant:\")\n",
    "            Rec = SeqRecord(Seq(sequence), id = name)\n",
    "            Rec.name = name\n",
    "            # Make a copy of the original, switch the fragments and add it to the list. \n",
    "            # Deep-copy ensures there are no pointer issues\n",
    "            tempVariant = copy.deepcopy(records)\n",
    "            tempVariant[toVary - 1] = Rec\n",
    "            variantRecords.append(copy.deepcopy(tempVariant))\n",
    "            print(\"\")\n",
    "\n",
    "    # Returns a list of lists of the SeqRecords of the fragments\n",
    "    return variantRecords\n",
    "\n",
    "def stitch(fragments):\n",
    "    #this function takes seq records and prints primers\n",
    "    \n",
    "    #let's make an empty sequence file\n",
    "    Nfrags=len(fragments)\n",
    "    donor=Seq(\"\")\n",
    "    index=[]\n",
    "    print(\"\")\n",
    "    for i in range (0, Nfrags):\n",
    "        donor=donor+fragments[i]\n",
    "    \n",
    "    for i in range (0, Nfrags):\n",
    "        if i==0:\n",
    "            print(\"Lup\"+ fragments[i].id + \" \" + getPrimer(donor))\n",
    "            print(\"Rup\"+ fragments[i].id + \"(\" + fragments[i+1].id + \") \" + overhangPrimer(fragments[i].reverse_complement(),fragments[i+1].reverse_complement()))\n",
    "        elif i==Nfrags-1:\n",
    "            print(\"Ldown\"+ fragments[i].id + \"(\" + fragments[i-1].id + \") \" + overhangPrimer(fragments[i],fragments[i-1]))\n",
    "            print(\"Rdown\"+ fragments[i].id + \" \" + getPrimer(donor.reverse_complement()))\n",
    "        else:\n",
    "            print(\"L\"+ fragments[i].id + \"(\" + fragments[i-1].id + \") \" + overhangPrimer(fragments[i],fragments[i-1]))\n",
    "            print(\"R\"+ fragments[i].id + \"(\" + fragments[i+1].id + \") \" + overhangPrimer(fragments[i].reverse_complement(),fragments[i+1].reverse_complement()))\n",
    "\n",
    "    print(\"\")\n",
    "    print(\"Your donor DNA cassette, has the following bp length and sequence:\")\n",
    "\n",
    "\n",
    "    print(\"\")\n",
    "    print(len(donor.seq))\n",
    "    print(\"\")\n",
    "\n",
    "    print(donor.seq)\n",
    "\n",
    "    print(\"\")\n",
    "    print(\"You might want to copy this entire prompt and save it for your records.\")\n",
    "        \n",
    "\n",
    "    \n",
    "def pickCut(Sequence):\n",
    "    from prettytable import from_csv\n",
    "\n",
    "    print(\"Which of these characterized loci is going to get lucky: \")\n",
    "\n",
    "    fp = open(\"cutsites.csv\", \"r\")\n",
    "    pt = from_csv(fp)\n",
    "    fp.close()\n",
    "\n",
    "    print(pt)\n",
    "\n",
    "    cutsiteNum = int(input(\"Choose your cutsite (Leftmost number):\"))\n",
    "\n",
    "    print(\"\")\n",
    "    print(\"You've chosen: \")\n",
    "    print(pt.get_string(start = cutsiteNum-1, end = cutsiteNum))\n",
    "\n",
    "    counter = 0\n",
    "    for row in pt:\n",
    "        counter += 1\n",
    "        if counter == cutsiteNum:\n",
    "            row.border = False\n",
    "            row.header = False\n",
    "            cutsite = row.get_string(fields=[\"sequence\"]).strip()\n",
    "            cutName = row.get_string(fields=[\"name\"]).strip()\n",
    "            chromosome = int(row.get_string(fields=[\"chromosome\"]).strip())\n",
    "    if chromosome >= 10:\n",
    "        filename = \"Scer\" + str(chromosome) + \".fasta\" \n",
    "    else:\n",
    "        filename = \"Scer0\" + str(chromosome) + \".fasta\"\n",
    "\n",
    "    ChromosomeRec=SeqIO.read(filename, \"fasta\")\n",
    "    ChromosomeSeq = ChromosomeRec.seq\n",
    "\n",
    "    if ChromosomeSeq.find(cutsite) == -1:\n",
    "        ChromosomeSeq = ChromosomeSeq.reverse_complement()\n",
    "\n",
    "    startInd = ChromosomeSeq.find(cutsite)\n",
    "\n",
    "    UpHomology = SeqRecord(ChromosomeSeq[startInd - 1000:startInd])\n",
    "    DownHomology = SeqRecord(ChromosomeSeq[startInd + 30:startInd + 1030])\n",
    "    return UpHomology, DownHomology\n",
    "\n",
    "\n",
    "\n",
    "#fetchGene(\"TEF1\")   \n",
    "#print (fetchNeighbor(\"TEF1\", \"downstream\", 250).seq)\n",
    "\n",
    "#print (buildCassette())\n",
    "askUser()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "        name  Chr          cut Sequence\n",
      "0    ARS106a    1  ATACGGTCAGGGTAGCGCCC\n",
      "1    ARS208a    2  GTCCGCTAAACAAAAGATCT\n",
      "2    ARS208b    2  TAGGATACAAGTGTCTGTGC\n",
      "3    ARS308a    3  CACTTGTCAAACAGAATATA\n",
      "4      LEU2a    3  AAAGGTGAGAGCGCCGGAAC\n",
      "5      MATαa    3  GCCCTGGATAGCGAAGCGTG\n",
      "6      MATαb    3  AATGATGTCTGGGTTTTGTT\n",
      "7      MATαc    3  TGGGAACAAGAGCAAGACGA\n",
      "8      RDS1a    3  ATTCAATACGAAATGTGTGC\n",
      "9    ARS416a    4  CCTTCAAAGGATTCACCAGA\n",
      "10   ARS416b    4  TTAGGGACAGTTAGAGGCGG\n",
      "11   ARS416c    4  CAATTCAGGCCAAATCTTGA\n",
      "12   ARS416d    4  TAGTGCACTTACCCCACGTT\n",
      "13   ARS511b    5  CAGTGTATGCCAGTCAGCCA\n",
      "14     URA3a    5  ATTCGCTATTATTCAAGCTT\n",
      "15     URA3b    5  CCTTCGTTCTTCCTTCTGCT\n",
      "16   ARS607b    6  CCAGCGTAAGGTAAATATTA\n",
      "17   ARS607c    6  CTATTTTTGCTTTCTGCACA\n",
      "18   SAP155b    6  GGTTTTCATACTGGGGCCGC\n",
      "19   SAP155c    6  ATGAAAGACAACTATAGGGC\n",
      "20     CAN1y    6  GATACGTTCTCTATGGAGGA\n",
      "21   ARS720a    7  CAACAATTGTTACAATAGTA\n",
      "22   ARS805a    8  TTATTTGAATGATATTTAGT\n",
      "23   ARS911b    9  GTAATATTGTCTTGTTTCCC\n",
      "24  ARS1021a   10  AAGTTGGCTAGCCATATTTT\n",
      "25  ARS1021b   10  CCTCTGTGTGGTGGTAATTG\n",
      "26  ARS1014a   10  TTATGTGCGTATTGCTTTCA\n",
      "27  ARS1114a   11  CTTGTGAAACAAATAATTGG\n",
      "28  ARS1206a   12  CGAACATTTTTCCATGCGCT\n",
      "29  ARS1309a   13  CCTGTGGTGACTACGTATCC\n",
      "30  ARS1414a   14  GCGCCACAGTTTCAAGGGTC\n",
      "31   YOLCd1a   15  CTAGAATTTCCATTTTGCGT\n",
      "32   YOLCd1b   15  GACTAGTTAAGCGAGCATGT\n",
      "33     HIS3b   15  AATATAGAGTGTACTAGAGG\n",
      "34  ARS1622a   16  TAAAGCCACCACATCGCAAA\n",
      "35  ARS1622b   16  GTCACGTTCCTGAGGTTACT\n",
      "36  YPRCd15a   16  ATATGTTTGGTTTCGATTGT\n",
      "37  YPRCd15c   16  AATCCGAACAACAGAGCATA\n",
      "38   ARS511a  NaN  GAATAAAGAATGATGATCGC\n",
      "39   SAP155a  NaN  ATGAAAGACAACCATAGGGC\n",
      "40   ARS601a  NaN  CCAGGTTGCCGCACGTGTGT\n",
      "41     HIS3a  NaN  ATTTTTTTTTTTCCACCTAG\n",
      "42   ARS911a  NaN  AGGGCAAATGATAAGTCGGC\n",
      "43  YPRCd15b  NaN  AAGTAACTCTACTCCGCCTG\n",
      "16.0\n"
     ]
    }
   ],
   "source": [
    "# import libraries we're using\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio import SeqIO\n",
    "from Bio.SeqUtils import MeltingTemp\n",
    "import copy\n",
    "\n",
    "from intermine.webservice import Service\n",
    "\n",
    "from pandas import *\n",
    "from pandas import DataFrame, read_csv\n",
    "import pandas as pd #this is how I usually import pandas\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "import sys\n",
    "\n",
    "\n",
    "df = pd.read_excel(\"locusTable.xlsx\")\n",
    "print(df)\n",
    "print(df['Chr'].max())\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.4.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
